### PR TITLE
Make ReadOnlyError on PROT_NONE consistent

### PR DESCRIPTION
**Pull request by Keno**
This replaces #36625. Essentially the question is what should
reading from PROT_NONE memory do. At the moment we throw
ReadOnlyMemoryError on Linux and pass through the segfault on
Windows/Mac. It's not super clear what the right answer is,
though it would be nice to be consistent. After some discussion,
it seemed that not throwing ReadOnlyMemoryError on Linux was probably
the better consistency direction for two reasons:

1. Getting a ReadOnlyMemoryError is confusing if the failing operation
   is a read. We could of course rename the error, but that would be
   a bigger breaking change.
2. The purpose of this exception is to protect people who accidentally
   try to write to mmap'ed memory that they don't have write permissions
   to. It was never intended to become a general memory access error
   exception, so it seems prudent to restrict it as much as possible.

To make this happen, there is some platform specific code here to
read the memory error register out of the signal context.
Decoding this register is shared code between the operating systems,
but the register itself is in an OS-specifc location in the signal
frame. I have implemented the retrieval code for all our CI'ed
platforms. Other users will get an appropriate `#warning` (similar
to other code in this file) and can fill in the necessary code
at their leisure. A new test in test/misc.jl will indicate whether
or not the check is correct.

### CODE REVIEW

File: src/signals-mach.c [hunk 1]
```diff
@@ -167,11 +161,11 @@ static void jl_call_in_state(jl_ptls_t ptls2, host_thread_state_t *state,
     uint64_t rsp = (uint64_t)ptls2->signal_stack + sig_stack_size;
     assert(rsp % 16 == 0);
 
+#ifdef _CPU_X86_64_
```
Comment (Keno) [inline src/signals-mach.c:164]: I meant to put this in a separate PR - oh well. It's correct anyway, and I'll follow up with the lr reset in a follow up PR.

**Comment by vchuravy** [comment]
> I have implemented the retrieval code for all our CI'ed platforms.

Except PPC ;) Where at least the build is now part of CI (https://build.julialang.org/#/builders/57/builds/1172)
How sure are you that the test is sufficient for this? https://build.julialang.org/#/builders/72/builds/1104/steps/5/logs/stdio has all the `misc` test passing.

**Comment by Keno** [comment]
> How sure are you that the test is sufficient for this? https://build.julialang.org/#/builders/72/builds/1104/steps/5/logs/stdio has all the `misc` test passing.

Less sure now that I've incorporated that bit of information.

**Comment by vchuravy** [comment]
> Less sure now that I've incorporated that bit of information.

Damn us Bayesian machines.

**Pull request merged**